<a href="https://colab.research.google.com/github/p09323028/2020a_NTU_Econometrics_I/blob/main/Homework/PS4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Homework 4**
Author: Jinze Wu

Student Number: p09323028

# **Excercise 10.1**
This exercise refers to the drunk driving panel data regression summarized in
Table 10.1.
1. New Jersey has a population of 8.85 million people. Suppose New Jersey
increased the tax on a case of beer by $2 (in 1988 dollars). Use the
results in column (5) to predict the number of lives that would be saved
over the next year. Construct a 99% confidence interval for your answer.
2. The drinking age in New Jersey is 21. Suppose that New Jersey lowered
its drinking age to 19. Use the results in column (5) to predict the change
in the number of traffic fatalities in the next year. Construct a 95% confidence
interval for your answer.
3. Suppose real income per capita in New Jersey increases by 3% in the
next year. Use the results in column (6) to predict the change in the
number of traffic fatalities in the next year. Construct a 95% confidence
interval for your answer.
4. How should standard errors be clustered in the regressions in columns
(2) through (7)?
5. How should minimum drinking age be included in the regressions?
Should it enter as a continuous variable or as a series of indicator variables?
Be specific about the information you use to assess this question.

---



![Table 10.1](http://drive.google.com/uc?id=1TGQkkEZ7avREKNJBLGaFTE2buQqoG3Jj)

## 10.1-1
$[2 \times (0.69)] \times 8.85 \times 10^6 \times 10^{-4} = 1221.3$

SE: $0.35 \times 2 \times 8.85 \times 10^2 = 619.5$

95\% 信賴區間: $[-374.5, 2817.1]$

## 10.1-2
$(0.08) \times 885 = 70.8$

SE: $0.07 \times 885 = 61.95$

95% 信賴區間: $[-50.6, 192.2]$

## 10.1-3
$[1.79 \times \frac{3}{100}] \times 8.85 \times 10^6 \times 10^{-4} = 47.52$

SE: $0.64 \times 0.03 \times 8.85 = 16.99$

95% 信賴區間: $[14.2, 80.8]$

## 10.1-4
由於各國的誤差項應具序列相關,
故應採 clustered standard error.

## 10.1-5
由下圖知最低年齡不具線性關係, 故採 indicator variables 較佳。

<img src="http://drive.google.com/uc?id=1at4WFnWWGOsQ6Cod-fToBc-bDkpoS8DL" width="200">

# **Excercise 10.4**
Using the regression in Equation (10.11), what are the slope and intercept for
1. Entity 1 in time period 1?
2. Entity 1 in time period 3?
3. Entity 3 in time period 1?
4. Entity 3 in time period 3?

---





$\text{Items}$ |$\text{Slpoe}$|$\text{Intercept}$
------------|:---:|:-------:
$i=1$, $t=1$|$\beta_1$|$\beta_0$
$i=1$, $t=3$|$\beta_1$|$\beta_0$
$i=3$, $t=1$|$\beta_1$|$\beta_0+\gamma_3$
$i=3$, $t=3$|$\beta_1$|$\beta_0+\gamma_3$



# **Excercise 10.8**
Consider observations $(Y_{it}, X_{it})$ from the linear panel data model
$$Y_{it} = X_{it}\beta_1 + \alpha_i + \lambda_{i}t + u_{it}$$
where $t = 1,...,T$; $i = 1,...,n$; and $a_i + \lambda_{i}t$ is an unobserved entity-specific time trend. How would you estimate $\beta_1$?

---



以 Dummy Variables 及時間 $t$ 估算,

即 $Y_{it} = X_{it}\beta_1 + [\beta_0 + D2_i + ... + Dn_i] + [\beta_2t+\delta_2(D2_i \times t)+\delta_n(Dn_i \times t)]$

如此, $\lambda_i = \beta_2 + \delta_i$

# **Excercise 10.11**
Let $\widehat{\beta}_1^{DM}$ denote the entity-demeaned estimator given in Equation (10.22), and let $\widehat{\beta}_1^{BA}$ denote the “before and after” estimator without an intercept, so that 
$\widehat{\beta}_1^{BA} = [\sum_{i=1}^n(X_{i2}-X_{i1})(Y_{i2}-Y_{i1})/\sum_{i=1}^n(X_{i2}-X_{i1})^2]$. Show that, if  $T = 2$, $\widehat{\beta}_1^{DM} = \widehat{\beta}_1^{BA}$. [Hint: Use the definition of $\widetilde{X}_{it}$ before Equation (10.22) to show that $\widetilde{X}_{i1} = -0.5(X_{i2} - X_{i1})$ and 
$\widetilde{X}_{i2} = 0.5(X_{i2} - X_{i1})$.]

---



$\widehat{\beta}^{DM}_1=\frac{\sum_{i=1}^n\sum_{t=1}^t\widetilde{X}_{it}\widetilde{Y}_{it}}{\sum_{i=1}^n\sum_{t=1}^T\widetilde{X}_{it}^2}$ [Equation (10.22)]

若 $t=2$, $\widetilde{X}_{it} = X_{it} - \frac{X_{i1}-X_{i2}}{2}$

$\widetilde{X}_{i1}=-0.5(X_{i2}-X_{i1})$

$\widetilde{X}_{i2}=0.5(X_{i2}-X_{i1})$

$\widehat{\beta}^{DM}_1=\frac{\sum_{i=1}^n\sum_{t=1}^2\widetilde{X}_{it}\widetilde{Y}_{it}}{\sum_{i=1}^n\sum_{t=1}^2\widetilde{X}_{it}^2}$

$=\frac{\sum_{i=1}^n[\frac{1}{4}(X_{i1}-X_{i2})(Y_{i1}-Y_{i2})+\frac{1}{4}(X_{i2}-X_{i1})(Y_{i2}-Y_{i1})]}{\sum_{i=1}^n[\frac{2}{4}(X_{i2}-X_{i1})^2]}$

$=\frac{\frac{2}{4}[\sum_{i=1}^n(X_{i2}-X_{i1})(Y_{i2}-Y_{i1})]}{\frac{2}{4}[\sum_{i=1}^n(X_{i2}-X_{i1})^2]}=\widehat{\beta}^{BA}_1$

# **E10.2**
Do citizens demand more democracy and political freedom as their incomes
grow? That is, is democracy a normal good? On the text website, **http://www.pearsonglobaleditions.com**, you will find the data file **Income_Democracy**,
which contains a panel data set from 195 countries for the years 1960, 1965, ...2000. A detailed description is given in **Income_Democracy_Description**,
available on the website. The data set contains an index of political freedom/
democracy for each country in each year, together with data on each country’s
income and various demographic controls. (The income and demographic controls
are lagged five years relative to the democracy index to allow time for
democracy to adjust to changes in these variables.)
1. Is the data set a balanced panel? Explain.
2. The index of political freedom/democracy is labeled $Dem\_ind$.
 - What are the minimum and maximum values of $Dem\_ind$ in the data
set? What are the mean and standard deviation of $Dem\_ind$ in the
data set? What are the 10th, 25th, 50th, 75th, and 90th percentiles of
its distribution?
 - What is the value of $Dem\_ind$ for the United States in 2000?
Averaged over all years in the data set?
 - What is the value of $Dem\_ind$ for Libya in 2000? Averaged over all
years in the data set?
 - List five countries with an average value of $Dem\_ind$ greater than 0.95; less than 0.10; and between 0.3 and 0.7.
3. The logarithm of per capita income is labeled $Log\_GDPPC$. Regress
$Dem\_ind$ on $Log\_GDPPC$. Use standard errors that are clustered by
country.
 - How large is the estimated coefficient on $Log\_GDPPC$? Is the
coefficient statistically significant?
 - If per capita income in a country increases by 20%, by how much is
$Dem\_ind$ predicted to increase? What is a 95% confidence interval
for the prediction? Is the predicted increase in $Dem\_ind$ large or
small? (Explain what you mean by large or small.)
 - Why is it important to use clustered standard errors for the regression?
Do the results change if you do not use clustered standard
errors?
4. - Suggest a variable that varies across countries but plausibly varies
little—or not at all—over time and that could cause omitted variable
bias in the regression in (c).
 - Estimate the regression in (c), allowing for country fixed effects.
How do your answers to (c)(i) and (c)(ii) change?
 - Exclude the data for Azerbaijan, and rerun the regression. Do the
results change? Why or why not?
 - Suggest a variable that varies over time but plausibly varies little—or
not at all—across countries and that could cause omitted variable
bias in the regression in (c).
 - Estimate the regression in (c), allowing for time and country fixed
effects. How do your answers to (c)(i) and (c)(ii) change?
 - There are additional demographic controls in the data set. Should
these variables be included in the regression? If so, how do the results
change when they are included?
5. Based on your analysis, what conclusions do you draw about the effects
of income on democracy?

---



下載資料

In [ ]:
!gdown --id '13jtuZRk9a1BbbZsydKtr-mVUGkn0RD8b' --output 'income_democracy.xlsx'

In [ ]:
!pip install linearmodels
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import linearmodels as plm

In [ ]:
inde = pd.read_excel('income_democracy.xlsx')
inde

country  year   dem_ind  log_gdppc  ...     age_5   educ  age_median  code
0      Andorra  1960       NaN        NaN  ...       NaN    NaN         NaN     1
1      Andorra  1965       NaN        NaN  ...       NaN    NaN         NaN     1
2      Andorra  1970  0.500000        NaN  ...       NaN    NaN         NaN     1
3      Andorra  1975       NaN        NaN  ...       NaN    NaN         NaN     1
4      Andorra  1980       NaN        NaN  ...       NaN    NaN         NaN     1
...        ...   ...       ...        ...  ...       ...    ...         ...   ...
1364  Zimbabwe  1980  0.666667   7.831994  ...  0.046867  2.147   15.600000   211
1365  Zimbabwe  1985  0.500000   7.876341  ...  0.047184  2.816   15.400000   211
1366  Zimbabwe  1990  0.166667   7.913739  ...  0.045149  2.828   16.000000   211
1367  Zimbabwe  1995  0.333333   7.977293  ...  0.045090  4.087   16.700001   211
1368  Zimbabwe  2000  0.166667   7.882274  ...  0.046125  4.428   17.000000   211

[1369 rows x 13 columns]

## E10.2-1
資料集裡, 國家有 195國, 年度有 9年 ,應共 1,755筆,

但裡頭僅 1,369筆, 故為 unbalanced panel.

## E10.2-2

### E10.2-2(i)

In [ ]:
inde.dem_ind.describe(percentiles=[0.1, 0.25, 0.5, 0.75, 0.9])

count    1266.000000
mean        0.499073
std         0.371337
min         0.000000
10%         0.000000
25%         0.166667
50%         0.500000
75%         0.833333
90%         1.000000
max         1.000000
Name: dem_ind, dtype: float64

### E10.2-2(ii)

In [ ]:
inde[(inde.country=='United States')&(inde.year==2000)]

country  year  dem_ind  ...    educ  age_median  code
1256  United States  2000      1.0  ...  12.179        34.0   192

[1 rows x 13 columns]

In [ ]:
inde[inde.country=='United States'].dem_ind.mean()

0.9855555560853746

### E10.2-2(iii)

In [ ]:
inde[(inde.country=='Libya')&(inde.year==2000)]

country  year  dem_ind  log_gdppc  ...     age_5  educ  age_median  code
717   Libya  2000      0.0        NaN  ...  0.049442   NaN        19.6   106

[1 rows x 13 columns]

In [ ]:
inde[inde.country=='Libya'].dem_ind.mean()

0.10925926764806122

### E10.2-2(iv)

In [ ]:
g = inde.groupby(['country']).mean().dem_ind

In [ ]:
g[g>0.95].head()

country
Australia    1.000000
Austria      0.993333
Barbados     1.000000
Belgium      1.000000
Belize       1.000000
Name: dem_ind, dtype: float64

In [ ]:
g[g<0.1].head()

country
Afghanistan    0.096667
Angola         0.066667
Brunei         0.055556
Burundi        0.047619
China          0.080370
Name: dem_ind, dtype: float64

In [ ]:
g[(g>0.3)&(g<0.7)].head()

country
Antigua       0.555556
Argentina     0.665926
Armenia       0.500000
Bangladesh    0.533333
Bolivia       0.551111
Name: dem_ind, dtype: float64

## E10.2-3

In [ ]:
df = inde.dropna(subset=['dem_ind','log_gdppc'])
df

In [ ]:
reg_10_2_3i = smf.ols(formula='dem_ind~log_gdppc', data=df)
result_10_2_3i = reg_10_2_3i.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result_10_2_3i.summary())

                            OLS Regression Results                            
Dep. Variable:                dem_ind   R-squared:                       0.438
Model:                            OLS   Adj. R-squared:                  0.438
Method:                 Least Squares   F-statistic:                     399.5
Date:                Mon, 05 Apr 2021   Prob (F-statistic):           5.22e-44
Time:                        06:39:23   Log-Likelihood:                -110.72
No. Observations:                 958   AIC:                             225.4
Df Residuals:                     956   BIC:                             235.2
Df Model:                           1                                         
Covariance Type:              cluster                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.3548      0.100    -13.544      0.0

### E10.2-3(i)
$\widehat{\beta}_{log\_gdppc} = 0.2357$

$Pvalue < 0.01$

統計上顯著異於0。

### E10.2-3(ii)

In [ ]:
test_10_2_3ii = ['0.2*log_gdppc=0']
result_10_2_3ii = result_10_2_3i.t_test(test_10_2_3ii)
print(result_10_2_3ii.summary(xname=['0.2*log_gdppc']))

                               Test for Constraints                              
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
0.2*log_gdppc     0.0471      0.002     19.987      0.000       0.043       0.052


$Dem\_ind$ 平均增加 $0.0471$;

95% 信賴區間: $[0.043, 0.052]$

$Dem\_ind$ 介於 0 至 1, 故相對 small.

### E10.2-3(iii)
因各國誤差項應具序列相關, 故應採 clusterted standard error.

若未採 clusterted standard error, $\widehat{\beta}_{log\_gdppc}$ 的標準誤將較 E10.2-3(i) 小。

## E10.2-4

### E10.2-4(i)
如各國風土民情。

### E10.2-4(ii)

In [ ]:
reg_10_2_4ii = smf.ols(formula='dem_ind~log_gdppc+C(country)', data=df)
result_10_2_4ii = reg_10_2_4ii.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result_10_2_4ii.summary())

                            OLS Regression Results                            
Dep. Variable:                dem_ind   R-squared:                       0.734
Model:                            OLS   Adj. R-squared:                  0.685
Method:                 Least Squares   F-statistic:                     26.63
Date:                Mon, 05 Apr 2021   Prob (F-statistic):           7.74e-07
Time:                        07:03:52   Log-Likelihood:                 247.99
No. Observations:                 958   AIC:                            -194.0
Df Residuals:                     807   BIC:                             540.6
Df Model:                         150                                         
Covariance Type:              cluster                                         
                                                   coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------------

/usr/local/lib/python3.7/dist-packages/statsmodels/base/model.py:1752: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 150, but rank is 1
  


$\widehat{\beta}_{log\_gdppc} = 0.0837$

$Pvalue < 0.01$

統計上顯著異於0。

In [ ]:
test_10_2_4ii = ['0.2*log_gdppc=0']
result_10_2_4ii = result_10_2_4ii.t_test(test_10_2_4ii)
print(result_10_2_4ii.summary(xname=['0.2*log_gdppc']))

                               Test for Constraints                              
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
0.2*log_gdppc     0.0167      0.006      2.667      0.008       0.004       0.029


95% 信賴區間: $[0.004, 0.029]$

### E10.2-4(iii)

In [ ]:
dff = df[df.country != 'Azerbaijan']
reg_10_2_4iii = smf.ols(formula='dem_ind~log_gdppc+C(country)', data=dff)
result_10_2_4iii = reg_10_2_4iii.fit(cov_type='cluster', cov_kwds={'groups':dff['country'], 'use_correction':False})
print(result_10_2_4iii.summary())

                            OLS Regression Results                            
Dep. Variable:                dem_ind   R-squared:                       0.734
Model:                            OLS   Adj. R-squared:                  0.685
Method:                 Least Squares   F-statistic:                     26.47
Date:                Mon, 05 Apr 2021   Prob (F-statistic):           8.37e-07
Time:                        07:09:47   Log-Likelihood:                 247.23
No. Observations:                 957   AIC:                            -194.5
Df Residuals:                     807   BIC:                             535.1
Df Model:                         149                                         
Covariance Type:              cluster                                         
                                                   coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------------

/usr/local/lib/python3.7/dist-packages/statsmodels/base/model.py:1752: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 149, but rank is 1
  


因 dataset 中, Azerbaijan 僅於 2000年有資料, 故不會改變結果。

### E10.2-4(iv)
如科技進步。

### E10.2-4(v)

In [ ]:
reg_10_2_4v = smf.ols(formula='dem_ind~log_gdppc+C(country)+C(year)', data=df)
result_10_2_4v = reg_10_2_4v.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result_10_2_4v.summary())

                            OLS Regression Results                            
Dep. Variable:                dem_ind   R-squared:                       0.761
Model:                            OLS   Adj. R-squared:                  0.714
Method:                 Least Squares   F-statistic:                     27.82
Date:                Mon, 05 Apr 2021   Prob (F-statistic):           8.61e-28
Time:                        07:15:59   Log-Likelihood:                 298.75
No. Observations:                 958   AIC:                            -279.5
Df Residuals:                     799   BIC:                             494.0
Df Model:                         158                                         
Covariance Type:              cluster                                         
                                                   coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------------

/usr/local/lib/python3.7/dist-packages/statsmodels/base/model.py:1752: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 158, but rank is 9
  


$\widehat{\beta}_{log\_gdppc} = 0.0536$

$Pvalue > 0.1$

統計上不顯著異於0。

In [ ]:
test_10_2_4v = ['0.2*log_gdppc=0']
result_10_2_4v = result_10_2_4v.t_test(test_10_2_4v)
print(result_10_2_4v.summary(xname=['0.2*log_gdppc']))

                               Test for Constraints                              
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
0.2*log_gdppc     0.0107      0.008      1.273      0.203      -0.006       0.027


95% 信賴區間: $[-0.006, 0.027]$

### E10.2-4(vi)

In [ ]:
dff = df.dropna()
dff.shape

(679, 13)

In [ ]:
reg_10_2_4vi = smf.ols(formula='dem_ind~log_gdppc+C(country)+C(year)+log_pop+age_2+age_3+age_4+age_5+educ', data=dff)
result_10_2_4vi = reg_10_2_4vi.fit(cov_type='cluster', cov_kwds={'groups':dff['country'], 'use_correction':False})
print(result_10_2_4vi.summary())

                            OLS Regression Results                            
Dep. Variable:                dem_ind   R-squared:                       0.734
Model:                            OLS   Adj. R-squared:                  0.683
Method:                 Least Squares   F-statistic:                     7.938
Date:                Mon, 05 Apr 2021   Prob (F-statistic):           7.64e-11
Time:                        07:21:48   Log-Likelihood:                 194.51
No. Observations:                 679   AIC:                            -171.0
Df Residuals:                     570   BIC:                             321.7
Df Model:                         108                                         
Covariance Type:              cluster                                         
                                             coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------

/usr/local/lib/python3.7/dist-packages/statsmodels/base/model.py:1752: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 108, but rank is 14
  


因仍可能有遺漏變數知問題, 故將 dataset 的變數引進, $log\_gdppc$ 效果更不顯著。

## E10.4-5
由前題知高 $log\_gdppc$, 未必有高 $dem\_ind$.

## 補充

In [ ]:
# Ordinary OLS
reg = smf.ols(formula='dem_ind~log_gdppc', data=df)
result = reg.fit(cov_type='HC1')
print(result.summary())

In [ ]:
# Cluster OLS
reg = smf.ols(formula='dem_ind~log_gdppc', data=df)
result = reg.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result.summary())

In [ ]:
# Cluster & Time Effect OLS
reg = smf.ols(formula='dem_ind~log_gdppc+C(year)', data=df)
result = reg.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result.summary())

In [ ]:
# Cluster & State Effect OLS
reg = smf.ols(formula='dem_ind~log_gdppc+C(country)', data=df)
result = reg.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result.summary())

In [ ]:
# Cluster & State Effect & Time Effect OLS
reg = smf.ols(formula='dem_ind~log_gdppc+C(country)+C(year)', data=df)
result = reg.fit(cov_type='cluster', cov_kwds={'groups':df['country'], 'use_correction':False})
print(result.summary())